pkg/util/policy/agent/agent.py

```diff
@@ -284,138 +258,231 @@ def make_new_full_ruleset(current_rules, new_rules):
+        rules[target_segment_forward_chain].insert(0,
+                _make_rule(target_segment_forward_chain, "-j %s" % policy_chain_name))
+
+        # Default rules per tenant.
+        if tenant_wide_policy_vector_chain not in rules:
+            rules[tenant_wide_policy_vector_chain] = []
+            rules[tenant_wide_policy_vector_chain].append(
+                _make_rule(tenant_wide_policy_vector_chain, '-m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT'))
+            rules[tenant_wide_policy_vector_chain].append(
+                _make_rule(tenant_wide_policy_vector_chain, '-m comment --comment POLICY_CHAIN_HEADER -j RETURN'))
+
+        # Loop over peers and fill top level policy chains with source matching rules
+        for peer in policy_def.get('peers'):
+            # Possible peers are
+            # CIDR - detected by cidr field
+            # peer - detected by peer type
+            # tid/sid - detected by tenant_network_id and segment_network_id
+            pr = peer.get('peer')
+            cidr = peer.get('cidr')
+            from_tenant = peer.get('tenant_network_id')
+            from_segment = peer.get('segment_network_id')
+
+            if pr:
+                if pr == "any":
+                    jump_rules = [ _make_rule(policy_chain_name, "-j %s") % in_chain_name ]
+                else:
+                    raise Exception("Unsupported value of peer %s" % pr)
+
+            elif cidr:
+                jump_rules = [ _make_rule(policy_chain_name, "-s %s -j %s") % (cidr, in_chain_name) ]
+
+            elif from_segment and from_tenant:
+                u32_in_match = _make_u32_match(addr_scheme, tenant, from_segment)
+                jump_rules = [
+                    _make_rule(policy_chain_name, '-m u32 --u32 "%s" -j %s' %
+                                                       (u32_in_match, in_chain_name))
+                ]
+
+            else:
+                raise Exception("Unknown peer type %s" % pr)
+
+            if not policy_chain_name in rules:
+                rules[policy_chain_name] = []
+                rules[policy_chain_name].append(
+                    _make_rule(policy_chain_name, '-m comment --comment PolicyId=%s -j RETURN' % policy_id))
+
+            for r in jump_rules:
+                rules[policy_chain_name].insert(0,r)
+
+    # Loop over rules and render protocol + action part of each rules
+    for r in _make_rules(policy_def.get('rules')):
+        for pc in policy_chains:
+            if not pc in rules:
+                rules[pc] = []
+        # and stuff copy of the rule into each policy chain
+        rules[pc].append(_make_rule(pc,r))
 
     return rules
 
+
+def _make_rules(policy_rules):
+    """
+    For each rules in policy_rules creates in/out rules in iptables_rules.
+
+    Returns updated list of iptables_rules
+    """
+    # For each rule in the policy create iptables rules.
+    for r in policy_rules:
+        if r['protocol'] == 'TCP':
```

Comment:
Do we want to be particular about the capitalization of the protocol? We could use ```r['protocol'].upper() == 'TCP'``` to be more generous about this.